### ISSUE TITLE
Electron has an executable stack

### ISSUE THREAD

**Issue by evandandrea** [electron/electron#11628]
* Electron version: 1.7.10
* Operating system: Ubuntu 16.04

Executable stacks are [problematic](https://wiki.ubuntu.com/SecurityTeam/Roadmap/ExecutableStacks). Electron applications distributed as [snaps](https://snapcraft.io) will encounter [AppArmor denials](https://forum.snapcraft.io/t/snap-and-executable-stacks/1812) because the prebuilt electron binary has an executable stack.

### Expected behavior

Electron does not have an executable stack.

### Actual behavior

Electron has an executable stack.

### How to reproduce

```
$ wget https://github.com/electron/electron/releases/download/v1.7.10/electron-v1.7.10-linux-x64.zip
$ unzip electron-v1.7.10-linux-x64.zip
$ execstack -q electron
X electron
```

**Comment by kewde** [comment]
Oh boy..

**Comment by jdstrand** [comment]
It is important to note that this is not just a "snap thing" and that removing the executable stack in electron builds will benefit all electron users, not just electron snap users. From https://forum.snapcraft.io/t/snap-and-executable-stacks/1812 (linked to by @evandandrea):

"Because code that has an executable stack is vulnerable to stack memory attacks, fixing the code itself will harden the code everywhere it is run, Ie, it will meaningfully improve the code outside of snappy; it just so happens it will also make the ‘m’ denial go away and allow the snap to work in strict mode."

**Comment by jdstrand** [comment]
@kewde - some electron snap publishers are choosing to workaround the issue with [snapcraft.yaml](https://forum.snapcraft.io/t/fonts-fail-to-load-when-desktop-plug-added/3414/3). That said, this bug is about fixing electron everywhere so it doesn't need an executable stack at all and it can therefore run more securely regardless of the packaging format.

**Comment by ckerr** [comment]
Confirmed for [1.7.10](https://github.com/electron/electron/releases/download/v1.7.10/electron-v1.7.10-linux-x64.zip):

```sh
$ execstack -q electron
X electron
$ file electron
electron: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.26, BuildID[sha1]=9a7e6ea5a9e957ea954386eb925d2cdf862e2de4, stripped
```

I'm _not_ seeing this when I build locally from master:

```sh
$ execstack -q ./out/D/electron
- ./out/D/electron
$ file ./out/D/electron
./out/D/electron: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, with debug_info, not stripped

```
So this may be a side-effect of the release build process.

**Comment by ckerr** [comment]
Update, when I do a full build, rather than a debug one, I'm still seeing this in master:

```sh
$ execstack ./out/D/electron
- ./out/D/electron
$ execstack ./out/R/electron
X ./out/R/electron
```

Investigating the cause...

**Comment by ckerr** [comment]
Not caused by any of the shared libraries that we link against:

```sh
ldd ./out/R/electron | perl -ane 'print $F[2],"\n";' | xargs execstack -q  | grep "^X"
```

returns nothing

**Comment by ckerr** [comment]
Appears to be coming from our build of libchromiumcontent and libwebkit

`ld` doesn't have a mechanism to warn for executable stacks, but `gold` does (`--warn-execstack`). By forcing a build with `gold` and by injecting the warning into our build:

```diff
diff --git a/brightray/brightray.gyp b/brightray/brightray.gyp
index efd2ba30a..49785adaa 100644
--- a/brightray/brightray.gyp
+++ b/brightray/brightray.gyp
@@ -46,9 +46,9 @@
           # unresolved symbols errors.
           'direct_dependent_settings': {
             'ldflags': [
-              '-Wl,--whole-archive',
+              '-Wl,--whole-archive,--warn-execstack',
               '<@(libchromiumcontent_libraries)',
-              '-Wl,--no-whole-archive',
+              '-Wl,--no-whole-archive,--warn-execstack',
             ],
           }
         }, {  # (Release build on Linux)
diff --git a/common.gypi b/common.gypi
index fff12d9b3..865066bb0 100644
--- a/common.gypi
+++ b/common.gypi
@@ -162,9 +162,9 @@
           ['OS=="linux" and libchromiumcontent_component==0', {
             # Prevent the linker from stripping symbols.
             'ldflags': [
-              '-Wl,--whole-archive',
+              '-Wl,--whole-archive,--warn-execstack',
               '<@(libchromiumcontent_v8_libraries)',
-              '-Wl,--no-whole-archive',
+              '-Wl,--no-whole-archive,--warn-execstack',
             ],
           }, {
             'libraries': [ '<@(libchromiumcontent_v8_libraries)' ],
diff --git a/script/build.py b/script/build.py
index 1a7378bf9..05580a5ce 100755
--- a/script/build.py
+++ b/script/build.py
@@ -26,7 +26,7 @@ def main():
   args = parse_args()
   for config in args.configuration:
     build_path = os.path.join('out', config[0])
-    ret = subprocess.call([ninja, '-C', build_path, args.target])
+    ret = subprocess.call([ninja, '-v', '-C', build_path, args.target])
     if ret != 0:
       sys.exit(ret)
```

...and by brute-forcing `/usr/bin/ld` to point to `/usr/bin/gold` _(not recommended, obviously, but useful for a quick and dirty test)_ I get this output:

```sh
[9/9] ../../vendor/llvm-build/Release+Asserts/bin/clang++ -Wl,--whole-archive,--warn-execstack /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libbase.a /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libwebrtc.a /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libnet.a /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libangle.a /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libskia.a /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libwebkit.a /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libmedia.a /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libcc.a /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libwebkitbindings.a /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libservices.a /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libpdfium.a /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libcomponents.a /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libppapi.a /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libwebkitmodules.a -Wl,--no-whole-archive,--warn-execstack -rpath \$ORIGIN -rdynamic -Wl,--export-dynamic -pthread --sysroot=/home/charles/src/electron/electron/vendor/debian_wheezy_amd64-sysroot -L/home/charles/src/electron/electron/vendor/debian_wheezy_amd64-sysroot/lib/x86_64-linux-gnu -Wl,-rpath-link=/home/charles/src/electron/electron/vendor/debian_wheezy_amd64-sysroot/lib/x86_64-linux-gnu -L/home/charles/src/electron/electron/vendor/debian_wheezy_amd64-sysroot/usr/lib/x86_64-linux-gnu -Wl,-rpath-link=/home/charles/src/electron/electron/vendor/debian_wheezy_amd64-sysroot/usr/lib/x86_64-linux-gnu -m64 -Wl,-O1 -Wl,--as-needed -Wl,--gc-sections -Wl,-rpath=\$ORIGIN/lib/ -Wl,-rpath-link=lib/ -o electron -Wl,--start-group obj/atom/app/electron.atom_main.o obj/libelectron_lib.a obj/brightray/libbrightray.a obj/vendor/breakpad/libbreakpad_client.a -Wl,--end-group lib/libnode.so -lpthread -lgtk-x11-2.0 -lgdk-x11-2.0 -lpangocairo-1.0 -latk-1.0 -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lpangoft2-1.0 -lpango-1.0 -lgobject-2.0 -lfreetype -lfontconfig -ldbus-1 -lX11-xcb -lxcb -lXi -lXcursor -lXdamage -lXrandr -lXcomposite -lXext -lXfixes -lXrender -lX11 -lXtst -lXss -lgconf-2 -lgmodule-2.0 -lrt -lglib-2.0 -lnss3 -lnssutil3 -lsmime3 -lssl3 -lplds4 -lplc4 -lnspr4 /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libffmpeg.so -lasound -lcap -lcups -ldl -lresolv -lexpat
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(p256-x86_64-asm.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(sha1-x86_64.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(md5-x86_64.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(ghash-x86_64.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(vpaes-x86_64.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(chacha20_poly1305_x86_64.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(x86_64-mont5.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(rdrand-x86_64.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(aesni-x86_64.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(sha256-x86_64.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(sha512-x86_64.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(aes-x86_64.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(x25519-asm-x86_64.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(aesni-gcm-x86_64.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(chacha-x86_64.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(rsaz-avx2.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(bsaes-x86_64.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libchromiumcontent.a(x86_64-mont.o): missing .note.GNU-stack section implies executable stack
/usr/bin/ld: warning: /home/charles/src/electron/electron/vendor/download/libchromiumcontent/static_library/libwebkit.a(SaveRegisters_x86.o): missing .note.GNU-stack section implies executable stack
```

**Comment by ckerr** [comment]
From looking at some of the sources associated with these object files, the pattern I'm seeing is these are all assembly files that don't specify GNU-stack behavior.

https://wiki.gentoo.org/wiki/Hardened/ has a good writeup of how to handle this. Using Occam's razor and _assuming for the moment (this needs to be confirmed)_ that these assembly files are simply being flagged as stack-executable because their authors omitted a `GNU-stack` note, we may want to submit patches upstream to correct this. This would be approprate for open source and would fix the issue at its root.

In the interim, we can also try compiling with `--noexecstack`. This has the dual benefit of (1) letting us test to see if the assembly code actually needs an executable stack or not, which would be a prerequisite before submitting upstream patches with `GNU-stack` notes and (2) solving the problem in the short term without having to wait for upstream. And indeed, this is probably the only workaround for versions of Electron that work with legacy copies of libchromiumcontent...

**Comment by flexiondotorg** [comment]
@ckerr Any update on this?

**Comment by develar** [comment]
electron-builder 20.0.7 produces correct snap (works on macOS also, since custom solution implemented in go is used).

**Comment by jkleinsc** [comment]
This issue has been resolved in the latest beta release, 2.0.0-beta.1.  It should also be fixed in the next 1.8.x release when #11955 is merged in.

**Comment by ckerr** [comment]
#11955 is merged so this will be in 1.8.3.

**Comment by amoonmm** [comment]
> Confirmed for [1.7.10](https://github.com/electron/electron/releases/download/v1.7.10/electron-v1.7.10-linux-x64.zip):
> 
> 
> 
> ```sh
> 
> $ execstack -q electron
> 
> X electron
> 
> $ file electron
> 
> electron: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.26, BuildID[sha1]=9a7e6ea5a9e957ea954386eb925d2cdf862e2de4, stripped
> 
> ```
> 
> 
> 
> I'm _not_ seeing this when I build locally from master:
> 
> 
> 
> ```sh
> 
> $ execstack -q ./out/D/electron
> 
> - ./out/D/electron
> 
> $ file ./out/D/electron
> 
> ./out/D/electron: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, with debug_info, not stripped
> 
> 
> 
> ```
> 
> So this may be a side-effect of the release build process.
